### PR TITLE
feat(etcd): add utility that samples persistent store (etcd) usage

### PR DESCRIPTION
**Pull request by tiagolobocastro**
This tool is used to sample persistent store (ETCD) usage at runtime from a live cluster.
By default, it makes use of the deployer library to create a local cluster running on docker.

### CODE REVIEW

File: utils/pstor-usage/src/pools.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+type NodeId = String;
+/// Resource manager for pools
+pub(crate) struct PoolMgr {
+    node_ids: Vec<NodeId>,
+    size_bytes: u64,
+    use_malloc: bool,
+}
+impl PoolMgr {
+    /// New `ResourceMgr` that creates pools with `size_bytes` in bytes
```
Comment (paulyoong) [inline utils/pstor-usage/src/pools.rs:17]: `ResourceMgr` -> `PoolMgr`?

File: utils/pstor-usage/src/resources.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    pub(crate) fn inner(&self) -> &Vec<Box<dyn ResourceSample>> {
+        &self.inner
+    }
+    /// Get length of the samples
+    pub(crate) fn len(&self) -> usize {
+        self.inner.len()
+    }
+    /// Add more samples
+    pub(crate) fn extend(&mut self, add: Vec<Box<dyn ResourceSample>>) {
+        self.inner.reverse();
```
Comment (paulyoong) [inline utils/pstor-usage/src/resources.rs:90]: Why do we need to reverse it and then reverse it again at the end of the function?

File: utils/pstor-usage/src/volumes.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
+use crate::resources::{ResourceDelete, ResourceMgr, ResourceSample};
+use openapi::{apis::Uuid, clients::tower::direct::ApiClient, models};
+
+/// Resource manager for volumes
+pub(crate) struct VolMgr {
+    n_replicas: u8,
+    size_bytes: u64,
+}
+impl VolMgr {
+    /// New `ResourceMgr` for volumes with `n_replicas` replicas and size in bytes
```
Comment (paulyoong) [inline utils/pstor-usage/src/volumes.rs:10]: `ResourceMgr` -> `VolMgr`?

**Review by paulyoong — APPROVED** [review]

File: utils/pstor-usage/src/etcd.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
+use crate::resources::{ResourceMgr, ResourceSample, ResourceSamples, Sampler};
+use etcd_client::StatusResponse;
+use openapi::{apis::Url, clients::tower::direct::ApiClient};
+
+#[derive(Clone)]
+pub(crate) struct Etcd(etcd_client::Client);
+impl Etcd {
+    pub(crate) async fn new(url: Url) -> anyhow::Result<Self> {
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply utils/pstor-usage/src/etcd.rs:8]: Agreed, done

File: utils/pstor-usage/src/pools.rs (hunk 1 above)
Comment (tiagolobocastro) [reply utils/pstor-usage/src/pools.rs:17]: But it does return a `ResourceMgr` ?

File: utils/pstor-usage/src/resources.rs (hunk 2 above)
Comment (tiagolobocastro) [reply utils/pstor-usage/src/resources.rs:90]: It's to keep the same order, we could probably do it in a simpler way though.. for now I've added a comment.

File: utils/pstor-usage/src/pools.rs (hunk 1 above)
Comment (paulyoong) [reply utils/pstor-usage/src/pools.rs]: Yeah, fair enough.

**Comment by tiagolobocastro** [comment]
bors merge

**Pull request merged**